### PR TITLE
Support ingestion of long/float dimensions

### PR DESCRIPTION
**Pull request by jon-wei**
This PR adds long and float implementations of DimensionHandler/DimensionIndexer/DimensionMerger, allowing ingestion of long/float typed dimensions.

This also changes the `EncodedTypeArray` type parameter in the interfaces above to `EncodedKeyComponentType`, removing the restriction that the individual fields of row keys during ingestion must be arrays (to allow for single numeric values in the keys, since long/floats don't support multivalue rows)

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/LongDimensionIndexer.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  @Override
+  public int getUnsortedEncodedKeyComponentHashCode(Long key)
+  {
+    return key.hashCode();
+  }
+
+  @Override
+  public Object convertUnsortedEncodedKeyComponentToActualArrayOrList(Long key, boolean asList)
+  {
+    return Lists.newArrayList(key);
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/LongDimensionIndexer.java]: ImmutableList.of may be cheaper. It just stores the `key` as a field.

File: processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  @Override
+  public Indexed<Float> getSortedIndexedValues()
+  {
+    return null;
+  }
+
+  @Override
+  public Float getMinValue()
+  {
+    return 0.0f;
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java]: Similar comment to min/max in the Long handler.

File: processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  @Override
+  public Float getUnsortedEncodedValueFromSorted(Float sortedIntermediateValue)
+  {
+    return sortedIntermediateValue;
+  }
+
+  @Override
+  public Indexed<Float> getSortedIndexedValues()
+  {
+    return null;
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java]: Similar comment to getSortedIndexedValues in the Long handler.

File: processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  @Override
+  public int getUnsortedEncodedKeyComponentHashCode(Float key)
+  {
+    return key.hashCode();
+  }
+
+  @Override
+  public Object convertUnsortedEncodedKeyComponentToActualArrayOrList(Float key, boolean asList)
+  {
+    return Lists.newArrayList(key);
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java]: Similar comment to convertUnsortedEncodedKeyComponentToActualArrayOrList in the Long handler.

File: docs/content/ingestion/index.md [hunk 5]
```diff
@@ -36,7 +36,21 @@ An example dataSchema is shown below:
         "format" : "auto"
       },
       "dimensionsSpec" : {
-        "dimensions": ["page","language","user","unpatrolled","newPage","robot","anonymous","namespace","continent","country","region","city"],
+        "dimensions": [
+          "page","language","user","unpatrolled","newPage","robot","anonymous","namespace","continent","country","region","city",
```
Comment (jon-wei) [inline docs/content/ingestion/index.md]: Moved these to new lines

File: docs/content/ingestion/index.md [hunk 6]
```diff
@@ -36,7 +36,21 @@ An example dataSchema is shown below:
...
+          {
+            "type": "LONG",
```
Comment (jon-wei) [inline docs/content/ingestion/index.md]: They can be lowercase, made them lowercase in the docs now

File: docs/content/ingestion/index.md [hunk 7]
```diff
@@ -169,10 +183,38 @@ handle all formatting decisions on their own, without using the ParseSpec.
 
 | Field | Type | Description | Required |
 |-------|------|-------------|----------|
-| dimensions | JSON String array | The names of the dimensions. If this is an empty array, Druid will treat all columns that are not timestamp or metric columns as dimension columns. | yes |
+| dimensions | JSON Object array | A list of [dimension schema](#dimension-schema) objects or dimension names. Providing a name is equivalent to providing a String-typed dimension schema with the given name. If this is an empty array, Druid will treat all columns that are not timestamp or metric columns as String-typed dimension columns. | yes |
```
Comment (jon-wei) [inline docs/content/ingestion/index.md]: Changed to "JSON array"

File: docs/content/ingestion/index.md [hunk 8]
```diff
@@ -169,10 +183,38 @@ handle all formatting decisions on their own, without using the ParseSpec.
...
 
+#### Dimension Schema
+A dimension schema specifies the type and name of a dimension to be ingested.
+
+For example, the following `dimensionsSpec` section from a `dataSchema` ingests one column as Long (`countryNum`), two columns as Float (`userLatitude`, `userLongitude`), and the other columns as Strings:
+
+```json
+"dimensionsSpec" : {
+  "dimensions": [
+    "page","language","user","unpatrolled","newPage","robot","anonymous","namespace","continent","country","region","city",
```
Comment (jon-wei) [inline docs/content/ingestion/index.md]: Moved dimensions to new lines

File: processing/src/main/java/io/druid/segment/LongDimensionIndexer.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+  @Override
+  public Long getUnsortedEncodedValueFromSorted(Long sortedIntermediateValue)
+  {
+    return sortedIntermediateValue;
+  }
+
+  @Override
+  public Indexed<Long> getSortedIndexedValues()
+  {
+    return null;
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/LongDimensionIndexer.java]: Changed this and the bitmap-related method to throw UnsupportedOperationException

File: processing/src/main/java/io/druid/segment/LongDimensionIndexer.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  @Override
+  public Long getMinValue()
+  {
+    return 0L;
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/LongDimensionIndexer.java]: Changed these to use MIN_VALUE and MAX_VALUE

File: processing/src/main/java/io/druid/segment/LongDimensionIndexer.java (hunk 1 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/LongDimensionIndexer.java]: Changed to use ImmutableList.of

File: processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java (hunk 3 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java]: Changed this and the bitmap-related method to throw UnsupportedOperationException

File: processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java (hunk 2 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java]: Changed to use min/max here as well

File: processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java (hunk 4 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java]: Changed to use ImmutableList.of

**Comment by jon-wei** [comment]
Re: your comment on makeColumnValueSelector

thanks, I missed that, I added a couple of tests, one in GroupBy and one in TopN that have aggregations on numeric columns ingested as dimensions, these trigger the float->long and long->float cases you mentioned in IncrementalIndexStorageAdapter

I added something to makeLong/FloatColumnSelector to create wrapping selectors that cast the values returned by the base selector from the indexer if necessary

**Comment by gianm** [comment]
@jon-wei The wrapping stuff in IISA feels a bit weird -- I think it'd be nicer for the DimensionIndexers to handle that on their own. So, that means replacing makeColumnValueSelector with makeObjectColumnSelector, makeLongColumnSelector, makeFloatColumnSelector, and makeDimensionSelector, giving full control over behavior to the type handler.

The long/float handlers should probably return NullDimensionSelector for makeDimensionSelector since QISA behaves like that too.

The string handler should probably return Zero-selectors for long/float since, again, QISA behaves like that.

**Comment by gianm** [comment]
Also, the current code in your patch throws UnsupportedOperationExceptions in some cases, but I think we should avoid those and return null/zero selectors instead.

**Comment by fjy** [comment]
👍

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java [hunk 11]
```diff
@@ -539,7 +546,70 @@ public int lookupId(String name)
   }
 
   @Override
-  public Object convertUnsortedEncodedArrayToActualArrayOrList(int[] key, boolean asList)
+  public LongColumnSelector makeLongColumnSelector(
+      IncrementalIndexStorageAdapter.EntryHolder currEntry, IncrementalIndex.DimensionDesc desc
+  )
+  {
+     return ZeroLongColumnSelector.instance();
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: How valid is it to return zero selector here? Maybe try to convert from String to number?

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java [hunk 12]
```diff
@@ -539,7 +546,70 @@ public int lookupId(String name)
...
+  }
+
+  @Override
+  public FloatColumnSelector makeFloatColumnSelector(
+      IncrementalIndexStorageAdapter.EntryHolder currEntry, IncrementalIndex.DimensionDesc desc
+  )
+  {
+    return ZeroFloatColumnSelector.instance();
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: Same

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 11 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: I believe QueryableIndexStorageAdapter does the same thing, so at least it's consistent with that. However, it'd be nice (for schema evolution of a column from string -> numeric) for the behavior to change to try to convert the string to a number.

File: processing/src/main/java/io/druid/segment/StringDimensionIndexer.java (hunk 12 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/StringDimensionIndexer.java]: Same as https://github.com/druid-io/druid/pull/3966#discussion_r125680562.

**Comment by licl2014** [comment]
@jon-wei @gianm  i have a doubt that long/float dimensions will cause high Cardinality  easily when we have a group by query.so have we consider this scenario?

**Comment by gianm** [comment]
@licl2014 it could happen, although in that case Druid will do the best it can. If cardinality is not super high it should be fine and go quickly. If cardinality is super high then Druid will spill to disk as needed (assuming disk spilling is enabled).

**Pull request merged**